components/org.apache.stratos.rest.endpoint/src/main/java/org/apache/stratos/rest/endpoint/api/StratosApiV41Utils.java

```diff
@@ -980,50 +981,41 @@ public static void addServiceGroup(GroupBean serviceGroupDefinition) throws Rest
 
-                findCartridgesInGroupBean(serviceGroupDefinition, cartridgeTypes);
+            //validate the group definition to check if cartridges duplicate in any groups defined
+            validateCartridgeDuplicationInGroupDefinition(serviceGroupDefinition);
 
-                Set<String> duplicates = findDuplicates(cartridgeTypes);
-                if (duplicates.size() > 0) {
-                    StringBuilder duplicatesOutput = new StringBuilder();
-                    for (String dup : duplicates) {
-                        duplicatesOutput.append(dup).append(" ");
-                    }
-                    if (log.isDebugEnabled()) {
-                        log.debug("duplicate cartridges defined: " + duplicatesOutput.toString());
-                    }
-                    throw new RestAPIException("Invalid Service Group definition, duplicate cartridges defined:" +
-                            duplicatesOutput.toString());
-                }
+            //validate the group definition to check if groups duplicate in any groups and
+            //validate the group definition to check for cyclic group behaviour
+            validateGroupDuplicationInGroupDefinition(serviceGroupDefinition, new ArrayList<String>());
 
-                CloudControllerServiceClient ccServiceClient = getCloudControllerServiceClient();
+            CloudControllerServiceClient ccServiceClient = getCloudControllerServiceClient();
 
-                cartridgeNames = new String[cartridgeTypes.size()];
-                int i = 0;
-                for (String cartridgeType : cartridgeTypes) {
-                    try {
-                        if (ccServiceClient.getCartridge(cartridgeType) == null) {
-                            // cartridge is not deployed, can't continue
-                            log.error("Invalid cartridge found in cartridge group " + cartridgeType);
-                            throw new RestAPIException("No Cartridge Definition found with type " + cartridgeType);
-                        } else {
-                            cartridgeNames[i] = cartridgeType;
-                            i++;
-                        }
-                    } catch (RemoteException e) {
-                        throw new RestAPIException(e);
-                    } catch (CloudControllerServiceCartridgeNotFoundExceptionException e) {
-                        throw new RestAPIException(e);
+            cartridgeNames = new String[cartridgeTypes.size()];
+            int j = 0;
+            for (String cartridgeType : cartridgeTypes) {
+                try {
+                    if (ccServiceClient.getCartridge(cartridgeType) == null) {
+                        // cartridge is not deployed, can't continue
+                        log.error("Invalid cartridge found in cartridge group " + cartridgeType);
+                        throw new RestAPIException("No Cartridge Definition found with type " + cartridgeType);
```

Comment:
We need to avoid throwing "RestAPIException" for specific issues. E.g. For "Invalid cartridge found in cartridge group", we need to throw an exception like, InvalidCartridgeFoundException" and catch it in StratosApiV41 class. Then we can send an appropriate status code according to error. We have few other places to fix as well.